File: bindings/python/ContinuousDynamicalSystem/tests/MultiStateWeightProvider.py
import pytest
pytestmark = pytest.mark.continuous_dynamical_system

import bipedal_locomotion_framework.bindings as blf
import numpy as np

def test_custom_task():
    dt = 0.01
    settling_time = 0.1
    tolerance = 1e-2

    walking_weight = np.array([10, 100, 1000])
    stance_weight = np.array([0, -10, 20])

    params = blf.parameters_handler.StdParametersHandler()
    params.set_parameter_float("settling_time", settling_time)
    params.set_parameter_float("sampling_time", dt)
    params.set_parameter_vector_string("states", ["WALKING", "STANCE"])

    params_walking = blf.parameters_handler.StdParametersHandler()
    params_stance = blf.parameters_handler.StdParametersHandler()

    params_walking.set_parameter_string("name", "walking")
    params_walking.set_parameter_vector_float("weight", walking_weight)

    params_stance.set_parameter_string("name", "stance")
    params_stance.set_parameter_vector_float("weight", stance_weight)

    assert params.set_group("WALKING", params_walking)
    assert params.set_group("STANCE", params_stance)

    provider = blf.continous_dynamical_system.MultiStateWeightProvider()
    assert provider.initialize(params)

    assert np.isclose(provider.get_weight(), walking_weight)
    assert provider.set_state("stance")

    for i in range(50):
        assert provider.advance()

    assert np.isclose(provider.get_weight(), stance_weight, rtol=tolerance)

    # reset the provider
    provider.reset("stance")
    assert np.isclose(provider.get_weight(), stance_weight)

    assert provider.set_state("walking")

    for i in range(50):
        assert provider.advance()

    assert np.isclose(provider.get_weight(), walking_weight, rtol=tolerance)

File: src/ContinuousDynamicalSystem/tests/MultiStateWeightProvider.cpp
/**
 * @file MultiStateWeightProvider.cpp
 * @authors Giulio Romualdi
 * @copyright 2022 Istituto Italiano di Tecnologia (IIT). This software may be modified and
 * distributed under the terms of the GNU Lesser General Public License v2.1 or any later version.
 */

#include <chrono>

// Catch2
#include <catch2/catch_test_macros.hpp>

#include <Eigen/Dense>

#include <BipedalLocomotion/ContinuousDynamicalSystem/MultiStateWeightProvider.h>
#include <BipedalLocomotion/ParametersHandler/StdImplementation.h>
#include <BipedalLocomotion/TestUtils/MemoryAllocationMonitor.h>

using namespace BipedalLocomotion::ContinuousDynamicalSystem;
using namespace BipedalLocomotion::TestUtils;

TEST_CASE("Multistate weight provider")
{
    using namespace std::chrono_literals;
    constexpr std::chrono::nanoseconds dT = 10ms;
    constexpr double settlingTime = 0.1;
    constexpr double tolerance = 1e-2;

    Eigen::Vector3d walkingWeight;
    walkingWeight << 10, 100, 1000;

    Eigen::Vector3d stanceWeight;
    stanceWeight << 50, 30, 20;

    auto params = std::make_shared<BipedalLocomotion::ParametersHandler::StdImplementation>();
    params->setParameter("settling_time", settlingTime);
    params->setParameter("sampling_time", dT);
    params->setParameter("states", std::vector<std::string>{"WALKING", "STANCE"});

    auto paramsWalking = std::make_shared<BipedalLocomotion::ParametersHandler::StdImplementation>();
    auto paramsStance = std::make_shared<BipedalLocomotion::ParametersHandler::StdImplementation>();

    paramsWalking->setParameter("name", "walking");
    paramsWalking->setParameter("weight", walkingWeight);

    paramsStance->setParameter("name", "stance");
    paramsStance->setParameter("weight", stanceWeight);

    REQUIRE(params->setGroup("WALKING", paramsWalking));
    REQUIRE(params->setGroup("STANCE", paramsStance));

    SECTION("Without reset")
    {
        MultiStateWeightProvider provider;
        REQUIRE(provider.initialize(params));

        REQUIRE(provider.getOutput().isApprox(walkingWeight));
        REQUIRE(provider.setState("stance"));
        REQUIRE_FALSE(provider.setState("foo"));

        for (unsigned int i = 0; i < 50; i++)
        {
            if (i >= 1)
            {
                MemoryAllocationMonitor::startMonitor();
            }
            REQUIRE(provider.advance());
            if (i >= 1)
            {
                REQUIRE(MemoryAllocationMonitor::endMonitorAndCheckNoMemoryAllocationInLastMonitor());
            }
        }

        REQUIRE(provider.getOutput().isApprox(stanceWeight, tolerance));
    }

    SECTION("With reset")
    {
        MultiStateWeightProvider provider;
        REQUIRE(provider.initialize(params));

        REQUIRE(provider.getOutput().isApprox(walkingWeight));
        REQUIRE(provider.setState("stance"));

        for (unsigned int i = 0; i < 50; i++)
        {
            if (i >= 1)
            {
                MemoryAllocationMonitor::startMonitor();
            }
            REQUIRE(provider.advance());
            if (i >= 1)
            {
                REQUIRE(MemoryAllocationMonitor::endMonitorAndCheckNoMemoryAllocationInLastMonitor());
            }
        }

        REQUIRE(provider.getOutput().isApprox(stanceWeight, tolerance));

        REQUIRE(provider.reset("stance"));
        REQUIRE(provider.getOutput().isApprox(stanceWeight));
        REQUIRE(provider.setState("walking"));
        for (unsigned int i = 0; i < 50; i++)
        {
            if (i >= 1)
            {
                MemoryAllocationMonitor::startMonitor();
            }
            REQUIRE(provider.advance());
            if (i >= 1)
            {
                REQUIRE(MemoryAllocationMonitor::endMonitorAndCheckNoMemoryAllocationInLastMonitor());
            }
        }

        REQUIRE(provider.getOutput().isApprox(walkingWeight, tolerance));
    }
}
